plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -356,6 +377,160 @@ private static ConnectorPageSource createOrcPageSource(
+            return Optional.empty();
+        }
+
+        Map<IcebergColumnHandle, Integer> originalColumnIndexes = Streams.mapWithIndex(columns.stream(), Map::entry)
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> toIntExact(entry.getValue())));
+        // Order the columns in path depth order so existing projections for nested columns can be reused
+        List<IcebergColumnHandle> orderedColumns = originalColumnIndexes.keySet().stream()
+                .sorted(Comparator.comparingInt(col -> col.getPath().size()))
+                .collect(toImmutableList());
+
+        ImmutableList.Builder<ColumnHandle> projectedColumns = ImmutableList.builder();
+        Integer[] outputColumnMapping = new Integer[columns.size()];
+        Map<List<String>, Integer> mappedColumnIndices = new HashMap<>();
+        int projectedColumnIndex = 0;
+
+        for (IcebergColumnHandle column : orderedColumns) {
+            Optional<Integer> existingProjectedColumn = findExistingProjection(mappedColumnIndices, column);
+            int columnIndex = originalColumnIndexes.get(column);
+            if (existingProjectedColumn.isPresent()) {
+                outputColumnMapping[columnIndex] = existingProjectedColumn.get();
+                continue;
+            }
+
+            projectedColumns.add(column.getBaseColumn());
+            mappedColumnIndices.put(column.getFullPath(), projectedColumnIndex);
+            outputColumnMapping[columnIndex] = projectedColumnIndex;
+            projectedColumnIndex++;
+        }
+
+        return Optional.of(new ReaderColumns(projectedColumns.build(), ImmutableList.copyOf(outputColumnMapping)));
+    }
+
+    private static Optional<Integer> findExistingProjection(Map<List<String>, Integer> mappedColumnIndices, IcebergColumnHandle column)
+    {
+        List<String> columnPath = column.getFullPath();
+        for (int prefixLength = columnPath.size(); prefixLength > 0; prefixLength--) {
+            List<String> prefix = columnPath.subList(0, prefixLength);
+            Integer existingColumnIndex = mappedColumnIndices.get(prefix);
+            if (existingColumnIndex != null) {
+                return Optional.of(existingColumnIndex);
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static Map<Integer, OrcColumn> mapIdsToOrcFileColumns(List<OrcColumn> columns)
+    {
+        ImmutableMap.Builder<Integer, OrcColumn> columnMap = ImmutableMap.builder();
+        columns.stream()
+                .filter(orcColumn -> orcColumn.getAttributes().containsKey(ORC_ICEBERG_ID_KEY))
+                .forEach(orcColumn -> {
+                    columnMap.put(getFieldId(orcColumn), orcColumn);
+                    columnMap.putAll(mapIdsToOrcFileColumns(orcColumn.getNestedColumns()));
+                });
+        return columnMap.build();
+    }
+
+    private static Map<Integer, org.apache.parquet.schema.Type> mapIdsToParquetType(List<org.apache.parquet.schema.Type> fields)
```

Comment:
I am too dumb for this. Please interleave the code with comments.

Reply:
Move it just below `projectColumnsForParquet` to keep parquet related code together.